pageserver/src/tenant/layer_map/layer_coverage.rs

```diff
@@ -87,9 +87,8 @@ impl<Value: Clone> LayerCoverage<Value> {
             }
             prev_covered = needs_cover;
         }
-        if !prev_covered {
-            to_remove.push(key.end);
-        }
+        // TODO check if the nodes inserted at key.start and key.end are safe
+        //      to remove. It's fine to keep them but they could be redundant.
```

Comment:
Should we add some tests for these which nodes are left in some conditions? This is a heuristics based idea, show that something works regardless of these extras which we currently don't remove, as I don't really catch up what they are and if they matter.

Or would it be the same as now added test with more asserts?

Reply:
They're not heuristics. These nodes are the points that describe the coverage line, almost like the points of a polyon. Polygons can have redundant points too. The only thing redundant points change is "what happens when you move other points". They stop being redundant then. They pin in place part of the curve you don't want to move.

That's the reason we add a node at `key.end` before doing anything else. So that when we mess with nodes in the layer's key range we don't change what happens after `key.end`.

The correct thing to do is to remove the node at the end if it's value is the same as the value at the previous node. Then there wouldn't leave any redundant nodes. But leaving it there works too. There won't be too many redundant nodes.

But I see there's some confusion with what nodes are. Is there a better name/docstring for what I'm describing here. Maybe `steps` (as in, staircase steps) instead of `nodes`? Let's work on this until the code/docstrings are clear. I don't want to be the only person that can debug it :)

Reply:
> They're not heuristics. These nodes ...

Sorry, I meant my question "Should we add some tests for these which nodes are left in some conditions?" was a heuristics based idea inspired by the TODO.

Having now read your polygon example it does start to make sense what these are. I do connect the polygon to the blog post on the subject but can't remember right away running into this in the comments. Looking around, perhaps this is somewhat described in `LayerCoverage::nodes`?

Reply:
Ok. Let me know if a rename/docstring/example would help somewhere

Reply:
Fun fact: Just realized that "depth buffer" and "z-buffer" are other synonyms for this concept of "coverage" https://en.wikipedia.org/wiki/Z-buffering

Reply:
While I know very little about gfx programming, the Z-buffer sounds like a good analogy.

If anything, we could include the brief mentions of your earlier polygons and Z-buffer near I guess `LayerCoverage`. The doc comment on the type is not huge. It might be natural to add a link to https://neon.tech/blog/persistent-structures-in-neons-wal-indexing at the same time.